src/java.base/share/classes/sun/security/ssl/NewSessionTicket.java

```diff
@@ -375,43 +362,104 @@ public byte[] produce(ConnectionContext context) throws IOException {
             // get a new session ID
             SSLSessionContextImpl sessionCache = (SSLSessionContextImpl)
                 hc.sslContext.engineGetServerSessionContext();
-            SessionId newId = new SessionId(true,
-                hc.sslContext.getSecureRandom());
+            int sessionTimeoutSeconds = sessionCache.getSessionTimeout();
+            if (sessionTimeoutSeconds > MAX_TICKET_LIFETIME) {
+                if (SSLLogger.isOn && SSLLogger.isOn("ssl,handshake")) {
+                    SSLLogger.fine("No session ticket produced: " +
+                            "session timeout");
+                }
+
+                return null;
+            }
 
             SecretKey resumptionMasterSecret =
                 hc.handshakeSession.getResumptionMasterSecret();
             if (resumptionMasterSecret == null) {
                 if (SSLLogger.isOn && SSLLogger.isOn("ssl,handshake")) {
-                    SSLLogger.fine(
-                            "No session ticket produced: " +
-                            "no resumption secret");
+                    SSLLogger.fine("No session ticket produced: " +
+                        "no resumption secret");
                 }
 
                 return null;
             }
 
-            // construct the PSK and handshake message
-            BigInteger nonce = hc.handshakeSession.incrTicketNonceCounter();
-            byte[] nonceArr = nonce.toByteArray();
-            SecretKey psk = derivePreSharedKey(
-                    hc.negotiatedCipherSuite.hashAlg,
-                    resumptionMasterSecret, nonceArr);
+            /*
+             * This thread addresses a Windows only networking issue found with
+             * SSLSocketBruteForceClose. A client that quickly closes after
```

Comment:
And you suggestion would be?

Reply:
This is a low level networking error beyond my control.  All this code can do is accept that the operating system has sent it a fatal error that has blocked the servers ability to read data from the socket on data that was by the client already.   This data is no lost, which is not a good situation to be in.  Catching the exception doesn't resolved the lost data.  A similar situation has occurred before with [JDK-8235973](https://bugs.openjdk.org/browse/JDK-8235973).  Their solution does not fit here as this is during a normal read operation, but shows working around the issue was necessary.

Reply:
On the contrary, you are in control of this error. The client OS resets the connection whenever the client closes the socket without reading all available data from the buffers. When the reset is delivered to the server, any data that was not received yet is lost.

The best approach depends on the type of traffic on the connection. If the client is expected to receive data, we can send the NewSessionTicket message as before. If we don't know if the client is expected to receive data, we should delay sending the NewSessionTicket messages until the server actually writes data over the connection.

Sending the NewSessionTicket messages in a thread only adds variability to the mix... without a thread, the messages were guaranteed to be sent before user data. Now the messages can be sent any time before, in the middle, or after user data.

OpenSSL added a function to configure the number of tickets sent automatically after the finished message, and a function to request sending a ticket with the next application data. We should probably do the same.

https://www.openssl.org/docs/manmaster/man3/SSL_new_session_ticket.html

Regarding the failing test, there are 2 options to fix it:
- configure the server to send zero tickets, or
- receive at least one byte of data on the client side before closing the socket.

Reply:
Maybe the comment could use some rewording, but I tried to not write whole bug report in a code comment.

When this was first prototyped as non-threaded, the SocketException occurred because multiple NSTs were sent after the Finished.  The client sent a few messages and closed the connection immediately.  The server, sending the NSTs, runs into the WIndows SocketException when it tries to read the data sent by the client.  Failing to receive the any data or the close_notify.  The variability of the thread allowed the OS to handle the reset correctly and allow the TLS server to receive the close_notify and the data that was sent.

Waiting for application data to cross the wire may not work for clients that connect and immediately start multiple sessions via resumption to transfer data.  It will not have any tickets to resume from.  Additionally delaying the NST may not mean this problem will go away if the client closes during the NST creation.

Reply:
I experimented with the code a little, and I didn't observe any failures after inlining the thread code. On the other hand, when I moved the `hc.handshakeOutput.flush();` line inside the loop that sends NewSessionTicket messages, I immediately got a test failure. I forgot that the SSLSocket actually tries to "deplete" the socket before closing it, pretty sure that's the reason why the test didn't fail.

Are you sure the thread is still necessary?

Reply:
The thread has been the best way to handle this so far, but I've tried other ways with mixed results.
As you may or may not have seen with your `hc.handshakeOutput.flush()` experiment, was all the NST writes can max out the packet buffer size, so it needs occasional flushing.
This is all about IO and not getting the Windows exception.  Writing a lot of handshake data at once is causing the problem.  Using the thread can allow the IO to breathe a bit, even with maybe a `sleep()` help spread out the load. 
I'm also examining a way to use the socket writing to distribute the NSTs with the flow of other data.  The spec doesn't require all of them be sent immediately, but making sent too slow risks no tickets for multi resume clients.
I feel the solution to this isn't going to be perfect and will take some compromises.